### ISSUE TITLE
Rapid causes epic hang, fails to download

### ISSUE THREAD

**Issue by mtroyka** [ZeroK-RTS/Zero-K-Infrastructure#558]
When using rapid manually (ie through the settings panel), double clicking any item on the left sometimes/often causes ZKL to freeze for minutes at a time before finally responding, during which time it gets disconnected from the server so that when the download finally starts it never progresses. When rapid is called automatically (ie for map downloads and zk updates) I've never seen this happen that I can recall.

For whatever reason this doesn't seem to happen nearly as badly when ZKL has just been started, which is the only way I got it to download anything at all.

**Comment by xponen** [comment]
could you share the content of your rapid?:
"your-datadir\packages\selected.list"

probably will help in replicating the bug.

**Comment by Licho1** [comment]
It's a known bug i changed the mode from async to sync when rewriting parts of code to use "await", due to some issues.

**Comment by xponen** [comment]
you mean "LoadMasterAndVersions()"?
we should return to Async and because only 1 code call it, we can just use  "do Thread.Sleep(); while(isRefreshed)" to get same behaviour as "await".

**Comment by Licho1** [comment]
Its not so easy, because packages are now not auto refreshed every 2 minutes, due to abma. Instead they are only checked when you request download, and download itself has to follow and function needs to return "Download" object to caller.
So you would have to block anyway..

*BUT* it's that way because its only temporary, the plan is to remove package refresh from client and move it to server (site) and store data in DB .. same stuff as I commented on your previous fix commit.

**Comment by xponen** [comment]
can I help abma more by increasing my spam timer from 3 second to 2 minute?

**Comment by Licho1** [comment]
You do it 3s but only when dl is requested, which should be ok for abma

**Comment by abma** [comment]
its not because of me, its because of https://github.com/ZeroK-RTS/Zero-K-Infrastructure/issues/198

**Comment by shaun0x00** [comment]
I don't know what you are attempting to fix.
ZKL is still opening > 10 HTTP connections simultaneously, and keeping them all open for way too long.

https://github.com/ZeroK-RTS/Zero-K-Infrastructure/blob/master/Shared/PlasmaDownloader/Packages/PackageDownloader.cs#L189

This logic is clearly wrong. Adding random delays won't fix it. You need to pipeline requests, you need to reuse connections, you need to close connections when they are no longer required.

If you are not ready to fix that, just remove the whole parallel thing... and pipeline everything.

**Comment by Licho1** [comment]
Specs say that each repo should be separate site/server, hence the parallel request makes perfect sense. You want to dl stuff asap and you don't know where it might be (because regexes were removed from master file), so you need to get response from all asap.

**Comment by abma** [comment]
> Specs say

which specs?

**Comment by Licho1** [comment]
Specs we designed when we were creating rapid with det. Not my fault springrts is abusing them. There is absolutely no reason to split multiple repos on same machine, that's not how it was meant to work! But all this is really meaningless, clients wont use rapid repo search for long.. it's "to be removed", they will get exact repo from plasma and dl that directly.

**Comment by abma** [comment]
> Specs we designed 

WHICH SPECS? Where are they written? there are none!

**Comment by Licho1** [comment]
I believe they were on caspring.org wiki. In original version, (first server was made in C# by me according to det) the master also had regex patterns so that correct repo could be found either from rapid tag or internal name and not all repos would have to be searched. This was later dropped in det's ML implementation due to simplicity but the idea of one repo per site was kept. Caspring had one and later springrts.com had the other one.

Don't really know when it became bastardized like this with bazilion of repos on same server and no regex, but it's not the way it was invented/intended and originally implemented...

**Comment by abma** [comment]
> I believe they were on caspring.org wiki.

i don't trust in believes. please give a link what you are refering to or they don't exist. even if they would exist, they would be more than 2 years old and a lot changed server side. with the change to git repos were split because it simplified the migration and split repos are a LOT easier to maintain. also its now possible to make a more smart client which fetches only  the needed versions.gz. and not a huge versions.gz. versions.gz became very large this is why a split makes a lot of sense: a single commit would make all clients to refetch the huge version.gz it, with the split only the smaller versions.gz is updated and downloaded.

what do you mean with regex?

**Comment by Licho1** [comment]
The master file, which lists repos, contained for each repo regex patterns for all possible "game" types stored in that repo. 

So if you searched zk:stable, you knew exactly from master that "zk" is stored in repo X and only requested versions from that repo.

**Comment by abma** [comment]
> So if you searched zk:stable, you knew exactly from master that "zk" is stored in repo X and only requested versions from that repo.

this information got lost: where is the code in zero-k which uses it?

**Comment by Licho1** [comment]
It was removed long ago before even the migration to git.

**Comment by abma** [comment]
so you are refering to specs which don't exist and code that doesn't exist as well?

**Comment by Licho1** [comment]
Technically it sure is somewhere in SVN  or git.

There are constant complains about problem that is not a real problem,
which exists only because rapid was steered of the intended path and
complains persist even after a fix that removed time checks and before
planned complete removal of this part of client.

So hell yeah, I think i can refer to long deleted code and specs.



2015-05-08 17:28 GMT+02:00 abma <notifications@github.com>:

**Comment by shaun0x00** [comment]
wow that was fast.
When implementing an HTTP client, what you should do is pool connections and enforce limits per server. A server is... an IP address. So, all the DNS requests for rapid repos resolve to the same IP address, that provide enough information for a well behaving HTTP client to actually limit the number of simultaneous connections.

Talking about specifications, I believe HTTP applies:
http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.1.4
RFC 2616 section 8.1.4 practical considerations

 Clients that use persistent connections SHOULD limit the number of simultaneous connections that they maintain to a given server. A single-user client SHOULD NOT maintain more than 2 connections with any server or proxy. A proxy SHOULD use up to 2*N connections to another server or proxy, where N is the number of simultaneously active users. These guidelines are intended to improve HTTP response times and avoid congestion.

**Comment by Licho1** [comment]
Seriously? Even damn browser uses many times more *to same IP* .. chrome 6,
IE 13, this is getting silly .. .


2015-05-08 18:26 GMT+02:00 shaun <notifications@github.com>:

**Comment by shaun0x00** [comment]
indeed, this is ridiculous. 

With my browser, connections are limited per server... as per HTTP specification.

It seems that, due to reality, bandwidth & cpu power has increased since RFC 2616 was published. So, the actual number used in reality, in my browser, is 6 connections per server.

https://stackoverflow.com/questions/8404464/increasing-google-chromes-max-connections-per-server-limit-to-more-then-6
http://trac.tools.ietf.org/wg/httpbis/trac/ticket/131

So licho, you are wasting everyone's time by trying to find flaws in bug reports instead of focusing on what actually caused the bug report to exist in the first place.

But knowing that, I'll help us bridge the next step.
What you can do now is argue that, unlike what I said previously, it seems the limit is per host not per IP. So you can still point out that it is a rapid design error, in using different domains for different repositories, in that it favors using many parallel connections. Excellent. Its a kind of nasty argument that you can make, that could even be considered valid. Except it doesn't solve anything... As there are still as many connections going to the server and kept open for too long.

Common sense should be enough...

**Comment by Licho1** [comment]
Wasting? Excuse me, but patches welcome! And how does hijacking ticket
about other issue and ranting solve things?
Dne 8. 5. 2015 22:37 napsal uživatel "shaun" <notifications@github.com>:

**Comment by mtroyka** [comment]
Erm, what exactly does the number of simultaneous connections or the architecture of the rapid network have to do with ZKL blocking when making a rapid request? And also, why does it only seem to block pathologically when using settings->rapid but not in battle rooms?

**Comment by Licho1** [comment]
It has absolutely nothing to do with blocking and block only happens in GUI because it there waits for the list. During standard download download runs on separate thread and won't block GUI.

**Comment by shaun0x00** [comment]
@mtroyka Like @Licho1 said, GUI blocks because its waiting on downloads.
What he doesn't understand is that the HTTP server may queue incoming requests, because there are too many already. This can cause the GUI thread to hang... and so, discussing how many connections are open and how they are (not) closed is appropriate.

https://webmasters.stackexchange.com/questions/26724/notification-for-too-many-apache-connection-errors-on-shared-hosting/26752#26752

**Comment by Licho1** [comment]
Really it has nothing to do with it, I explicitly changed the code to blocking, knowing perfectly well what i'm doing!
It's changed to blocking, because parts of underlying lower level code changed to C# specific await/async pairs which handle thread switching and the top level code not.
So you need non async func to wait for async code, which can cause deadlock, if it's executed from main GUI thread (our case).

So I changed the code to blocking, to avoid deadlock, knowing it will only block rapid GUI and nothing else..

Sigh..

**Comment by shaun0x00** [comment]
See issue #614 for a source of leaked connections.

Connections are limited per host, by the framework (or some layer in it)

See https://msdn.microsoft.com/en-us/library/system.net.servicepoint%28v=vs.110%29.aspx
https://msdn.microsoft.com/en-us/library/system.net.servicepoint.connectionlimit%28v=vs.110%29.aspx

If you leak connections somewhere, then, opening many simultaneous connections in a blocking environment can only be more likely to block. @Licho1, what is happening when ServicePoint.ConnectionLimit is reached? Is it blocking the call to Create() the webrequest or does it fail with an exception?

Since there are not timeouts in ZK network code, it is not able to detect that something is wrong, and just hangs.

**Comment by Licho1** [comment]
fixed now
